### ISSUE TITLE
Seeding stats functions

### ISSUE THREAD

**Issue by maichmueller** [sympy/sympy#20522]
Hello,

since sampling in 1.7 has changed to support scipy as default, how can one seed their distributions and especially reseed dynamically?

Before (i.e. pre 1.7) the dists relied on the python module `random` by default and with `random.seed(int)` one could (clumsily) seed the rng machine and therefore create reproducibility. Seeing as how scipy is now used, is there a way to forward e.g. a numpy random state generator (or similar) to the `sample_iter` or the RVs themselves to provide them with the current state machine?
I couldn't find anything related to reproducibility in the docs and the code didnt reveal something useful either.

Cheers,
Michael

**Comment by oscarbenjamin** [comment]
CC @czgdp1807 @Smit-create 

Thanks for report Michael.

In a broader sense it seems that sympy needs to have a documented API for seeding which is perhaps a bigger problem. See also #12628 and #16967 which are about sympy's internal use of the random module.

For now though I guess the fact that random.seed no longer works could be viewed as a regression to be fixed in a 1.7.1 release. I don't know the sampling code so well but maybe a simple fix could be getting random.seed and using it to seed scipy and then passing the seed from scipy back to the random module afterwards.

**Comment by czgdp1807** [comment]
IMHO, messing up with seeds inside `sympy.stats` would create a lot of confusion. 
Whatever library is being used, seed should be set according to it's API. `sympy.stats` will generate samples using that library, so reproducibility won't be an issue. Otherwise as said above, there will be a lot conflicts between global seed settings and manipulations done by `sympy.stats`. For example, for `scipy`, the following code produces the same results when the seed is set is using `numpy`. In fact, `scipy` also uses(externally) `numpy.random.seed()` to set it's seed.

```python
from __future__ import division
from sympy import *
x, y, z, t = symbols('x y z t')
k, m, n = symbols('k m n', integer=True)
f, g, h = symbols('f g h', cls=Function)
init_printing()

from sympy.stats import *
import numpy
numpy.random.seed(seed=23)

N = Normal('N', 0, 1)
itr = sample(N, numsamples=1000)
for i in range(10):
    print(next(itr))
```
Run 1
```python
0.0433729598719631
1.616090294284517
0.72397598084618
-0.5757388849891719
-0.7686674942515839
0.4851699029666263
-0.9655320196808684
-0.2729587391435831
0.3003695264316943
-0.22258284437658313
```

Run - 2
```python
0.0433729598719631
1.616090294284517
0.72397598084618
-0.5757388849891719
-0.7686674942515839
0.4851699029666263
-0.9655320196808684
-0.2729587391435831
0.3003695264316943
-0.22258284437658313
```


> For now though I guess the fact that random.seed no longer works could be viewed as a regression to be fixed in a 1.7.1 release.

I don't think this is a thing that should be fixed as there is nothing wrong with this. `random.seed` is for setting seeds for python. For different libraries the ways will be different. Seed manipulation shouldn't be done by `sympy.stats` to avoid conflicts between the settings made by the end user and the changes in it because of manipulations done by `sympy.stats`.

> random.seed and using it to seed scipy

As shown above, `scipy` uses, `numpy.random.seed` so `random.seed` won't work.

What we can do on our part is to add additional message in the warning that seed should be set externally according to the external library used.

**Comment by czgdp1807** [comment]
> with `random.seed(int)` one could (clumsily) seed the rng machine and therefore create reproducibility.

That is the good way actually. In fact, I would prefer to avoid any manipulations done to seed settings internally by `sympy.stats`. If there are currently any, they should be removed.

**Comment by oscarbenjamin** [comment]
The best kind of seeding API is if it can be an explicit function parameter giving the user the option to avoid any dependency on global state.

> As shown above, `scipy` uses, `numpy.random.seed` so `random.seed` won't work.

My point was that sympy could *get* the random seed from random and use it to *set* the random seed for numpy. In general that does not seem a good approach but it is a response to the fact that previously the only way to control this from the user's perspective was with random.seed. That means that in some sense this is a regression compared to expected behaviour before.

**Comment by czgdp1807** [comment]
The API has been changed for sampling. So probably we can add some additional messages to set the seed externally. 
Allowing or not allowing seed is quite debatable. Sometimes calls to sympy.stats APIs may be just a part of larger code where the seeds have been set globally. Now passing the seed again to our APIs would be quite redundant. That is why I proposed to allow no manipulations with the seeds internally and let the user be aware of their settings enabling more transparency.

**Comment by oscarbenjamin** [comment]
The normal way to do this is to make the seed an optional argument:
```
def sample(seed=None):
    if seed is None:
        seed = get_global_seed()
    ...
```
No one needs to provide the seed if they don't want to. If they do want to then they can see exactly where to do it.

**Comment by maichmueller** [comment]
Relying on global seeding is never the appropriate way to seed a method imho. Suppose you use Sympy in conjunction with other libraries all seeding individually via numpy.random.seed, the end result is a mess in reproducibility because of these interferences (and a nightmare to track and debug I can say from experience). Precisely because there may be other parties influencing the global seed, there should be the option to go for a local rng state generator as numpy also supports (and recommends if memory serves well).
In fact, when I first noticed that sympy needed global seeding I was reluctant to use it for any statistical packages, since this is so easily breakable.

I agree with @oscarbenjamin to allow the seed to be passed as an optional parameter. Even more so I would recommend the seed parameter to accept a numpy generator next to an int.

**Comment by czgdp1807** [comment]
Well, now if sympy manipulates the seed internally and then it has to reset it again back to the global settings if there were any. Obtaining the global seed can be hard. Even though it may be doable, it will create a lot of complications and inconsistencies for the user, i.e., setting, resetting the seed and then returning the results. Between the setting and restting of seeds other global settings will be useless. 
Though still adding support for internal seed manipulation won't be possible before `1.7.1` because it will be needing a lot of code changes because every library will be having it's own way of setting the seed and we would be required to make changes everywhere. @Upabjojr Any thoughts on this?

**Comment by czgdp1807** [comment]
https://github.com/scipy/scipy/issues/3975 is a similar issue. We don't know if such features are there for other libraries. Though we can forward the received seed from the user to libraries which allow custom seeds to their distribution. But still it will take some time for this to be done.

**Comment by oscarbenjamin** [comment]
> But still it will take some time for this to be done.

Yes, but if sampling is worth doing then it's worth doing properly. SymPy is often used in scientific research and not providing reliable seed reproducibility is a major weakness for that. It looks like scipy fixed that issue in 2014. In SymPy we can fix this one in 2020. We also need #16967. Fixing these properly is a bit much for a bugfix release though.

The question right now is: given that random.seed is the way to seed sympy's sampling methods in 1.6.2 and no longer works in 1.7 should that be restored somehow in a bugfix 1.7.1 right now?

(We already need a 1.7.1 because of the sampling regression in #20521 anyway)

**Comment by czgdp1807** [comment]
> given that random.seed is the way to seed sympy's sampling methods in 1.6.2 and no longer works in 1.7 should that be restored somehow in a bugfix 1.7.1 right now?

I don't think, because that would mean undoing the support for external libraries because, `random.seed` works only for `python`. For `scipy` and `numpy`, `numpy.random.seed` should be used. I am not sure about `pymc3`.

**Comment by czgdp1807** [comment]
Though if needed, an additional parameter can be added for seed(an `int`) in the higher level API and forwarded to the respective libraries wherever possible. It will take some time(`1.7.1` may be a delayed if review period is longer) but the issue will be permanently fixed. That's the best that can be done about this.

**Comment by Upabjojr** [comment]
I would just let the seed be handled by the external libraries. SymPy is not meant to handle pseudorandom number generators.

**Comment by Upabjojr** [comment]
> Though if needed, an additional parameter can be added for seed(an `int`) in the higher level API and forwarded to the respective libraries wherever possible.

Well, that would cause the problem of handling the seeds with different backends. I would not modify the current sampling API.

**Comment by oscarbenjamin** [comment]
> I would just let the seed be handled by the external libraries. SymPy is not meant to handle pseudorandom number generators.

The logical conclusion from this is that all of the sampling code should be removed.

**Comment by Upabjojr** [comment]
> > I would just let the seed be handled by the external libraries. SymPy is not meant to handle pseudorandom number generators.
> 
> The logical conclusion from this is that all of the sampling code should be removed.

No, you can simply call:
```python
numpy.random.seed(42)
sample(Normal("N", 1, 0), 1000)
```

Using the `numpy` seed will set the seed for NumPy and SciPy (if those are the chosen backends).

**Comment by Upabjojr** [comment]
The problem is that seeding may depend on the chosen backend. For example, we may add more sampling libraries for random variables in the future. Those may have different ways of seeding their RNGs.

**Comment by czgdp1807** [comment]
Though I thought of one fix after some research. Like `scipy.random_variable.rvs` has `random_state` parameter which accepts integer seed values. Numpy has RandomState object through which we can sample using custom seed values provided. And `pymc3` has `rseed`. But again for evey library there is a different way to provide seeds, so it would be complicated to manage support for custom seeds for all of these. 
I think the major issue is with multi-processing architectures with each process using different seed where global seeds will pose a problem. But `sympy` isn't meant for such high compute purposes, though if it is then we can have the pain of managing different backends. 
There is a tradeoff here between consistency and simplicity of API. So, basically rather than taking extreme steps like coming up with a permanent but difficult fix or just doing away with all the sampling code, what we can do is open an issue and just wait if it receives enough support for allowing seed as an extra parameter and if it does then we can go with the fix of managing different backends for forwarding seed values to them.

**Comment by Upabjojr** [comment]
Another option would be to add the backends as code generators instead of samplers. In that case, given a SymPy expression, users can generate the code used to generate a random sample using an external library. How to manage the seed is up to the user then.

**Comment by oscarbenjamin** [comment]
> The problem is that seeding may depend on the chosen backend. For example, we may add more sampling libraries for random variables in the future. Those may have different ways of seeding their RNGs.

They will have different ways of seeding their RNGs however most libraries:

1. Allow setting the seed with an integer
2. Provide `set_state/get_state` methods

SymPy should make use of those. SymPy's `sample_iter` function has a `library=` argument. We can have `get_state/set_state` functions that also have a `library` argument. The state objects don't need to be interchangeable across different libraries. SymPy can also have a `seed` function that takes an integer. That can be used to make a local RNG from the random module which can in turn be used to `seed()` any libraries if a seed  is not provided.

Even if sympy does turn sampling into code generation we will still want to provide users with a way of getting random samples. That should be done without altering the global seed states of the different libraries which means

1. Using a local PRNG
2. Allowing a way to seed that PRNG

A library that provides random sampling without a way to set the seed and without the promise not to alter the global seed is deficient.

I agree entirely that random sampling is a whole can of worms that is best left to other libraries. I think it's fine that sympy provides an interface to those other libraries without attempting to reimplement them. The minimum standards for that interface are not currently met though.

**Comment by czgdp1807** [comment]
Currently, there are three libraries supported for externally sampling handmade and pre-defined distributions,

1. `scipy` - It is done using `.rvs` method which has `random_state` argument where we can pass the seed(an `int`) received from the end user. Minor changes in code will be required to do that.
2. `numpy` - A local `numpy.RandomState` object can be created which will receive the seed from the end user and generate samples according to it.
3. `pymc3` - It has a `rseed` argument for each of it's distributions.

So, for each new library added we will be needing to pass the seed from the end user. This would help in instance specific seeds and help in avoiding shared seeds across various instances. It will also not change the global seed settings of the end user. So, everything will be good.

There two drawbacks of this of this approach, instead of using global seed settings are as follows,

1. If by chance seed receives wrong values or may be due to a bug instead of `None` some wrong integer is passed, then it will create a mess while debugging as it will be hard to figure what happened to the seed inside sympy.stats.
2. Pain of managing different APIs for each library.

I am opening a PR against `1.7` for more details. If people like it then we may merge this directly to `1.7`, otherwise the changes can always go to `master`. 
Though for now the only quick fix to @maichmueller issue is to use `numpy.random.seed` instead of `random.seed`. If my PR gets merged then users will be able to pass `seed` as an optional argument.

**Comment by maichmueller** [comment]
Thanks for the great discussion so far!

To add my own two cents (although @oscarbenjamin already covered some of this): I'm unaware as to the reason why sympy supports three different libraries for sampling and I think this may the main crux of the situation. Sympy is basically trying to cover ground which apparently is not covered by each lib individually. Wouldn't a better solution be to support one library (e.g. Scipy) and contribute to this lib when needing further distributions instead of adding more libraries? While this is certainly a slower process and not as "independent", it would be the proper way. This would also solve directly how to set the seed.

**Comment by czgdp1807** [comment]
I think supporting extra libraries will be good. Say there is someone interested in sampling using GPU(e.g., when they are training bayesian neural networks by providing custom handmade priors) then when tensorflow support will be added, they will be able to do that with the help of `sympy.stats`. I was also thinking to add support for `cupy`(GPU counter part of `numpy`). The point is that it is always good to have more tools in the toolbox.

**Comment by maichmueller** [comment]
While I do agree that having more tools in the box is certainly good, the question is one of 'would be cool to have' vs 'people have to be able to use our pkg for this'. 
I think your GPU example is valid, since it includes a different domain of sampling uncovered by the other libraries. And if sympy should be used for this, then it needs to be added. However, e.g. scipy and numpy cover the same sampling logic and one literally builds on the other. So maybe numpy is actually not needed?

**Comment by oscarbenjamin** [comment]
I'm sure that anyone who was using this to train bayesian networks in a serious way would want to be able to record the seed used at the start of the training process and would want set/get functions for reproducibility. 

Personally I find it hard to imagine someone in that scenario wanting to use sympy for sampling though. They could just use the underlying sampling libraries directly and they would probably want to if they are so concerned about rng performance that it needs to run on the GPU.

**Comment by czgdp1807** [comment]
Well, whatever libraries are added are added, in future libraries will be added only when there will be convincing use cases. 


> They could just use the underlying sampling libraries directly and they would probably want to if they are so concerned about rng performance that it needs to run on the GPU.

Well, training BNN was just an example. In fact no one can perfectly predict what people may or may not do. The combination of custom RVs and sampling certainly is helpful IMHO.

> to record the seed used at the start of the training process and would want set/get functions for reproducibility

Well after #20528 anyone would be able to pass the seed as an integer value and reproduce results easily. The difference made is that seeds are localized to distributions and neither fully depend nor manipulate the environment settings.

**Comment by maichmueller** [comment]
Since the seed parameter has been added, I will be closing this issue.
